webapp/packages/plugin-data-spreadsheet-new/src/DataGrid/DataGridSelection/useGridSelectionContext.tsx

```diff
@@ -9,93 +9,159 @@
+      const columnsLength = tableData.columns?.length || 0;
 
-    const columnsLength = (modelResultData.data as SqlResultSet).columns?.length || 0;
+      if (!multiple) {
+        selectedCells.clear();
+      }
 
-    if (!multiple) {
-      selectedCells.clear();
-    }
+      const rowSelection = [];
+      for (let i = 0; i < columnsLength; i++) {
+        rowSelection.push(i);
+      }
 
-    const rowSelection = [];
-    for (let i = 0; i < columnsLength; i++) {
-      rowSelection.push(i);
-    }
+      const firstRowIdx = Math.min(startPosition, lastPosition);
+      const lastRowIdx = Math.max(startPosition, lastPosition);
 
-    selectedCells.set(rowIdx, rowSelection);
-  }), [selectedCells, modelResultData]);
+      for (let rowIdx = firstRowIdx; rowIdx <= lastRowIdx; rowIdx++) {
+        if (isRowSelected(rowIdx)) {
+          selectedCells.delete(rowIdx);
+          continue;
+        }
 
-  const isRowSelected = useCallback((rowIdx: number) => {
-    if (!modelResultData) {
-      throw new Error('Model result data must be provided');
-    }
+        if (temporary) {
+          temporarySelectedCells.set(rowIdx, rowSelection);
+        } else {
+          selectedCells.set(rowIdx, rowSelection);
+        }
+      }
+    }), [selectedCells, temporarySelectedCells]);
 
-    const columnsLength = (modelResultData.data as SqlResultSet).columns?.length || 0;
+  const isRowSelected = useCallback((rowIdx: number) => {
+    const columnsLength = tableData.columns?.length;
 
     return selectedCells.get(rowIdx)?.length === columnsLength;
-  }, [modelResultData, selectedCells]);
+  }, [tableData, selectedCells]);
 
-  const unSelectRow = useCallback((rowIdx: number) => {
-    selectedCells.delete(rowIdx);
-  }, [selectedCells]);
+  const selectRange = useCallback(action(
+    (startPosition: number, lastPosition: number, columns: number[], multiple: boolean, temporary = false) => {
+      if (!multiple) {
+        selectedCells.clear();
+      }
+
+      const top = Math.min(startPosition, lastPosition);
+      const bottom = Math.max(startPosition, lastPosition);
+
+      for (let rowIdx = top; rowIdx <= bottom; rowIdx++) {
+        if (temporary) {
+          temporarySelectedCells.set(rowIdx, [...columns]);
+        } else {
+          const currentRowSelection = selectedCells.get(rowIdx) || [];
+          const newRowSelection = [...currentRowSelection, ...columns]
+            .filter((columnIdx, idx, arr) => arr.indexOf(columnIdx) === idx);
+
+          selectedCells.set(rowIdx, newRowSelection);
+        }
+      }
+    }), [selectedCells, temporarySelectedCells]);
+
+  const updateMultiSelection = useCallback(
+    (startPosition: IPosition, lastPosition: IPosition, multiple: boolean, temporary: boolean) => {
+      const columnsInRange = tableData.getColumnsInRange(startPosition.idx, lastPosition.idx);
+      const isIndexColumnInRange = tableData.isIndexColumnInRange(columnsInRange);
+
+      temporarySelectedCells.clear();
```

Comment:
`temporarySelectedCells.clear();` move to `selectRows` and `selectRange` functions, other move to callback